### PR TITLE
Follow up from abstract bn PR

### PR DESCRIPTION
**Pull request by nouiz**
I merged https://github.com/Theano/Theano/pull/5190 too fast. Here are some updates for corner cases.

I still need to locally make sure all tests pass and that it don't make the tests too long (also in DebugMode) before merging this PR.

The master isn't in a bad state. Just that corner case won't be as optimized as they should and some tests crash when floatX=float64.

### CODE REVIEW

**Comment by slefrancois** [comment]
jenkins test this

File: theano/gpuarray/dnn.py [hunk 1]
```diff
@@ -3037,7 +3032,7 @@ def local_log_softmax_dnn(node):
             len(node.inputs[0].clients) == 1):
         if version(raises=False) < 3000:
```
Comment (abergeron) [inline theano/gpuarray/dnn.py]: We don't support cudnn older than v4 anyway so this entire check could go away.

File: theano/gpuarray/tests/test_dnn.py [hunk 2]
```diff
@@ -26,6 +26,11 @@
 
 from theano.configdefaults import SUPPORTED_DNN_CONV_ALGO_FWD
 
+mode_with_gpu = mode_with_gpu.including()
+mode_without_gpu = mode_without_gpu.including()
+mode_with_gpu.check_py_code = False
+mode_without_gpu.check_py_code = False
```
Comment (abergeron) [inline theano/gpuarray/tests/test_dnn.py]: If you want to change mode_without_gpu globally to not check python code, that would be ok.

File: theano/sandbox/cuda/dnn.py [hunk 3]
```diff
@@ -3705,6 +3705,19 @@ def local_gpu_batch_norm_inference_inplace(node):
                                          inplace=True)(*node.inputs)]
 
 
+def values_eq_approx_high_tol(a, b):
+    """
+    This fct is needed to don't have DebugMode raise useless
+    error due to ronding error.
```
Comment (abergeron) [inline theano/sandbox/cuda/dnn.py]: error -> errors (the first word, not the second).
ronding -> rounding

File: theano/sandbox/cuda/dnn.py [hunk 4]
```diff
@@ -3705,6 +3705,19 @@ def local_gpu_batch_norm_inference_inplace(node):
...
+
+    This happen as We reduce on the two last dimensions, so this
```
Comment (abergeron) [inline theano/sandbox/cuda/dnn.py]: We -> we

File: theano/sandbox/cuda/dnn.py [hunk 5]
```diff
@@ -3705,6 +3705,19 @@ def local_gpu_batch_norm_inference_inplace(node):
...
+    can raise the absolute error if the number of element we
```
Comment (abergeron) [inline theano/sandbox/cuda/dnn.py]: element -> elements

File: theano/tensor/nnet/bn.py [hunk 6]
```diff
@@ -413,15 +413,27 @@ def infer_shape(self, node, shape):
     def make_node(self, x, scale, bias, epsilon=1e-4,
                   running_average_factor=0.1,
                   running_mean=None, running_var=None):
+        x = as_tensor_variable(x)
+        scale = as_tensor_variable(scale)
+        bias = as_tensor_variable(bias)
+        epsilon = as_tensor_variable(epsilon)
```
Comment (abergeron) [inline theano/tensor/nnet/bn.py]: This can cause trouble with floatX=float32 and epsilon=1e-5.  I don't know if we want to address those.

File: theano/tensor/nnet/tests/test_bn.py [hunk 7]
```diff
@@ -201,7 +201,7 @@ def test_batch_normalization_train():
                                               bn.AbstractBatchNormTrainGrad))
                             for n in f.maker.fgraph.toposort()])
             # run
-            for data_shape in ((5, 10, 30, 40, 10), (4, 3, 1, 1, 1), (1, 1, 5, 5, 5)):
```
Comment (abergeron) [inline theano/tensor/nnet/tests/test_bn.py]: You don't need to change the first shape here?

File: theano/tensor/type.py [hunk 8]
```diff
@@ -204,8 +204,8 @@ def filter_variable(self, other, allow_convert=True):
         Convert a symbolic Variable into a TensorType, if compatible.
 
         For the moment, only a TensorType or CudaNdarrayType will be
```
Comment (abergeron) [inline theano/tensor/type.py]: While you're there, GpuArrayType will convert here too.  As will scalars, I think.

File: theano/tensor/nnet/bn.py (hunk 6 above)
Comment (nouiz) [inline theano/tensor/nnet/bn.py]: I think we want the normal floatX behavior here too. If epsilon don't have a dtype, we use the same algo. I prefer to not introduce special case about our floatX behavior.

File: theano/tensor/nnet/tests/test_bn.py (hunk 7 above)
Comment (nouiz) [inline theano/tensor/nnet/tests/test_bn.py]: Not for the CPU tests. I only fixed those that I need to.

File: theano/tensor/type.py (hunk 8 above)
Comment (nouiz) [inline theano/tensor/type.py]: Not scalar for now:

$ python -c "import theano;s=theano.scalar.float32(); ts=theano.tensor.scalar(); ts.type.filter_variable(s)"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "theano/tensor/type.py", line 234, in filter_variable
    self=self))
TypeError: Cannot convert Type float32 (of Variable <float32>) into Type TensorType(float64, scalar). You can try to manually convert <float32> into a TensorType(float64, scalar).


But maybe we should. Anyway, too close to an rc to change that now. I'll update about GpuArrayType.

File: theano/gpuarray/tests/test_dnn.py [hunk 9]
```diff
@@ -26,6 +26,10 @@
 
 from theano.configdefaults import SUPPORTED_DNN_CONV_ALGO_FWD
 
+mode_with_gpu = mode_with_gpu.including()
+# Globally disabled for mode_without_gpu
+mode_with_gpu.check_py_code = False
```
Comment (abergeron) [inline theano/gpuarray/tests/test_dnn.py]: What I meant by globally was in config.py directly, but this is fine too.

Comment (nouiz) [inline theano/gpuarray/tests/test_dnn.py]: I don't like doing it globally. I'm pretty sure many tests rely on that to check the correspondance of both code. But maybe doing the test in FAST_COMPILE as we do do enough, but I'm not 100% sure of that.
